src/java.rs

```diff
@@ -0,0 +1,224 @@
+use crate::mapper::ProguardMapper;
+
+fn java_base_types(encoded_ty: char) -> Option<&'static str> {
+    match encoded_ty {
+        'Z' => Some("boolean"),
+        'B' => Some("byte"),
+        'C' => Some("char"),
+        'S' => Some("short"),
+        'I' => Some("int"),
+        'J' => Some("long"),
+        'F' => Some("float"),
+        'D' => Some("double"),
+        'V' => Some("void"),
+        _ => None,
+    }
+}
+
+fn byte_code_type_to_java_type(byte_code_type: &str, mapper: &ProguardMapper) -> Option<String> {
+    let mut chrs = byte_code_type.chars();
+    let token = chrs.next()?;
+    if token == 'L' {
+        // invalid signature
+        let l = chrs.clone().last();
+        if l.is_none() || l.unwrap() != ';' {
+            return None;
+        }
+        chrs.next_back(); // remove final `;`
+        let obfuscated = chrs.collect::<String>().replace('/', ".");
+
+        if let Some(mapped) = mapper.remap_class(&obfuscated) {
+            return Some(mapped.to_string());
+        }
+
+        return Some(obfuscated);
+    } else if token == '[' {
+        let type_sig = chrs.clone().collect::<String>();
+        if !type_sig.is_empty() {
+            return Some(format!(
+                "{}[]",
+                byte_code_type_to_java_type(chrs.collect::<String>().as_str(), mapper)
```

Comment:
```suggestion
                byte_code_type_to_java_type(&type_sig, mapper)
```

But if you use `char_indices` as suggested above, that would already be a `&str`.

The other thing to note here is that `byte_code_type_to_java_type` is recursive here, and the only reason it needs to return a `String` is because of these arrays.
It would be nice to remove the recursion instead of a manual loop maybe.
Depends on how frequent such types are.